tls/s2n_server_key_exchange.c

```diff
@@ -29,164 +30,133 @@
+    return 0;
+}
+
+int s2n_dhe_server_recv_params(struct s2n_connection *conn, struct s2n_blob *data_to_verify)
 {
     struct s2n_stuffer *in = &conn->handshake.io;
-    struct s2n_blob p, g, Ys, serverDHparams, signature;
     uint16_t p_length;
     uint16_t g_length;
     uint16_t Ys_length;
-    uint16_t signature_length;
 
     /* Keep a copy to the start of the whole structure for the signature check */
-    serverDHparams.data = s2n_stuffer_raw_read(in, 0);
-    notnull_check(serverDHparams.data);
+    data_to_verify->data = s2n_stuffer_raw_read(in, 0);
+    notnull_check(data_to_verify->data);
 
     /* Read each of the three elements in */
     GUARD(s2n_stuffer_read_uint16(in, &p_length));
-    p.size = p_length;
-    p.data = s2n_stuffer_raw_read(in, p.size);
+    struct s2n_blob p = { .size = p_length, .data = s2n_stuffer_raw_read(in, p_length)};
     notnull_check(p.data);
 
     GUARD(s2n_stuffer_read_uint16(in, &g_length));
-    g.size = g_length;
-    g.data = s2n_stuffer_raw_read(in, g.size);
+    struct s2n_blob g = {.size = g_length, .data = s2n_stuffer_raw_read(in, g_length)};
     notnull_check(g.data);
 
     GUARD(s2n_stuffer_read_uint16(in, &Ys_length));
-    Ys.size = Ys_length;
-    Ys.data = s2n_stuffer_raw_read(in, Ys.size);
+    struct s2n_blob Ys = {.size = Ys_length, .data = s2n_stuffer_raw_read(in, Ys_length)};
     notnull_check(Ys.data);
 
     /* Now we know the total size of the structure */
-    serverDHparams.size = 2 + p_length + 2 + g_length + 2 + Ys_length;
-
-    if (conn->actual_protocol_version == S2N_TLS12) {
-        s2n_hash_algorithm hash_algorithm;
-        s2n_signature_algorithm signature_algorithm;
-        GUARD(s2n_get_signature_hash_pair_if_supported(in, &hash_algorithm, &signature_algorithm));
-
-        GUARD(s2n_hash_init(&conn->secure.signature_hash, hash_algorithm));
-    } else {
-        GUARD(s2n_hash_init(&conn->secure.signature_hash, conn->secure.conn_hash_alg));
-    }
-
-    GUARD(s2n_hash_update(&conn->secure.signature_hash, conn->secure.client_random, S2N_TLS_RANDOM_DATA_LEN));
-    GUARD(s2n_hash_update(&conn->secure.signature_hash, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN));
-    GUARD(s2n_hash_update(&conn->secure.signature_hash, serverDHparams.data, serverDHparams.size));
-
-    GUARD(s2n_stuffer_read_uint16(in, &signature_length));
-    signature.size = signature_length;
-    signature.data = s2n_stuffer_raw_read(in, signature.size);
-    notnull_check(signature.data);
-
-    gt_check(signature_length, 0);
-
-    S2N_ERROR_IF(s2n_pkey_verify(&conn->secure.server_public_key, &conn->secure.signature_hash, &signature) < 0, S2N_ERR_BAD_MESSAGE);
-
-    /* We don't need the key any more, so free it */
-    GUARD(s2n_pkey_free(&conn->secure.server_public_key));
+    data_to_verify->size = 2 + p_length + 2 + g_length + 2 + Ys_length;
 
     /* Copy the DH details */
     GUARD(s2n_dh_p_g_Ys_to_dh_params(&conn->secure.server_dh_params, &p, &g, &Ys));
```

Comment:
Do we care that this is now running before signature verification instead of after it? There's a lot happening inside s2n_dh_p_g_Ys_to_dh_params and I wonder if this leaks timing information.

Reply:
Discussed in https://github.com/awslabs/s2n/issues/918 and fixed with the two new function in s2n_kex: read_data and parse_data.